compiler/rustc_codegen_llvm/src/debuginfo/metadata.rs

```diff
@@ -454,9 +454,9 @@ pub fn type_di_node<'ll, 'tcx>(cx: &CodegenCx<'ll, 'tcx>, t: Ty<'tcx>) -> &'ll D
         ty::RawPtr(ty::TypeAndMut { ty: pointee_type, .. }) | ty::Ref(_, pointee_type, _) => {
             build_pointer_or_reference_di_node(cx, t, pointee_type, unique_type_id)
         }
-        // Box<T, A> may have a non-1-ZST allocator A. In that case, we
-        // cannot treat Box<T, A> as just an owned alias of `*mut T`.
-        ty::Adt(def, args) if def.is_box() && cx.layout_of(args.type_at(1)).is_1zst() => {
+        // Only "native" Boxes (with the global allocator) are treated as equivalent to raw
+        // pointers.
+        ty::Adt(..) if t.is_box_global(cx.tcx) => {
             build_pointer_or_reference_di_node(cx, t, t.boxed_ty(), unique_type_id)
         }
```

Comment:
This is just debug info, the previous check is ok imo.

Reply:
Checking for a 1-ZST seems entirely random. I'd rather not have some parts of the compiler do one check and some parts do a different check.

Reply:
It's not random, it's effectively checking whether `Box<T, A>` is a newtype of `*mut T`.

Not a semantic equivalence, just a representational one. Ideally it shouldn't even look at the type parameters, but perhaps compare `box_layout.abi` with `box_layout.field(0).abi`?

The only exception would be if this is generating debuginfo specific to the old `~T`, i.e. language-special owned pointer, in which case @RalfJung would be right, tho it's not obvious that's what's happening.

Reply:
Maybe we should make the pointer field type an assoc type on allocator, then allocators can choose if they want unique or raw pointer behaviour

Reply:
> It's not random, it's effectively checking whether Box<T, A> is a newtype of *mut T.

It's still inconsistent. Since it's only debug info it doesn't seem worth introducing a new distinction aside from "Box with Global allocator" and "Box with custom allocator".

Reply:
Ideally it should just hit some case where `ty::Adt`s that happen to be pointer newtypes (perhaps checking for `#[repr(transparent)]` as well?) are special-cased.

Or if what actually happens by default is debuginfo being recorded as `struct` types (and newtypes just happen to be the same size as one of their fields), that shouldn't break anything, it would just look a bit awkward to have to deal with the fields.
Honestly, this kind of usecase really wants "smart pointer debuginfo", with the actual layout/ABI potentially allowing anything around the wrapped pointer (doesn't C++'s `unique_ptr` have the same issue? I wonder what they do for debuginfo).

Anyway I don't particularly care about debuginfo blocking anything, it's generally best-effort, if there's no obvious middle ground you could open an issue (if one doesn't already exist) about the general case of `Box`/`Rc`/`Arc`etc. and making sure we're not much worse than e.g. C++ `unique_ptr`/`shared_ptr`, even with custom allocators (on both sides).